File: sdk/samples/FrameworkSamples/NCL/SecureStreams/cs/negotiateserver/negotiateserver.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.IO;
using System.Net;
using System.Text;
using System.Net.Sockets;
using System.Net.Security;
using System.Security.Authentication;
using System.Security.Principal;

namespace Microsoft.Samples.SecureStreams
{
    static class NegotiateServer
    {
        public static void Main()
        {
            NegotiateStream negotiateStream = null;
            try
            {
                // Set the TcpListener on port 13000.
                Int32 port = 13000;
                IPAddress localAddr = IPAddress.Loopback;

                // TcpListener server = new TcpListener(port);
                TcpListener server = new TcpListener(localAddr, port);

                // Start listening for client requests.
                server.Start();

                // Buffer for reading data
                Byte[] bytes = new Byte[256];
                String data = null;

                // Enter the listening loop.
                while (true)
                {
                    Console.Write("Waiting for a connection... ");

                    // Perform a blocking call to accept requests.
                    // You could also user server.AcceptSocket() here.
                    TcpClient client = server.AcceptTcpClient();
                    Console.WriteLine("Connected!");

                    data = null;

                    // Get a stream object for reading and writing
                    // Wrap it in a NegotiateStream.
                    negotiateStream = new NegotiateStream(client.GetStream());
                    negotiateStream.AuthenticateAsServer();

                    if (negotiateStream.IsAuthenticated)
                    {
                        Console.WriteLine(
                            "IsAuthenticated: {0}",
                            negotiateStream.IsAuthenticated);
                        Console.WriteLine(
                            "IsMutuallyAuthenticated: {0}",
                            negotiateStream.IsMutuallyAuthenticated);
                        Console.WriteLine(
                            "IsEncrypted: {0}",
                            negotiateStream.IsEncrypted);
                        Console.WriteLine(
                            "IsSigned: {0}",
                            negotiateStream.IsSigned);
                        Console.WriteLine(
                            "IsServer: {0}",
                            negotiateStream.IsServer);
                        IIdentity remoteIdentity =
                            negotiateStream.RemoteIdentity;
                        Console.WriteLine(
                            "Client identity: {0}",
                            remoteIdentity.Name);
                        Console.WriteLine(
                            "Authentication Type: {0}",
                            remoteIdentity.AuthenticationType);
                    }

                    int i;

                    // Loop to receive all the data sent by the client.
                    while (
                        (i = negotiateStream.Read(bytes, 0, bytes.Length)) 
                        != 0)
                    {
                        // Translate data bytes to a string.
						// The encoding used is application specific.
                        data = 
                            System.Text.Encoding.ASCII.GetString(bytes, 0, i);
                        Console.WriteLine("Received: {0}", data);

                        // Process the data sent by the client.
                        data = data.ToUpper(
                            System.Globalization.CultureInfo.CurrentCulture);

                        byte[] msg = System.Text.Encoding.ASCII.GetBytes(data);

                        // Send back a response.
                        negotiateStream.Write(msg, 0, msg.Length);
                        Console.WriteLine("Sent: {0}", data);
                    }
                }
            }
            catch (AuthenticationException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (SocketException ex)
            {
                Console.WriteLine(ex.Message);
            }
            catch (IOException ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                if (negotiateStream != null)
                    negotiateStream.Close();
            }

            Console.WriteLine("\nHit enter to continue...");
            Console.Read();
        }
    }

}

File: sdk/samples/FrameworkSamples/NCL/SecureStreams/vb/negotiateserver/negotiateserver.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Net.Security
Imports System.Security.Authentication
Imports System.Security.Principal

Module NegotiateServer

    Sub Main()
        Dim negotiateStream As NegotiateStream = Nothing
        Try
            ' Set the TcpListener on port 13000.
            Dim port As Integer = 13000
            Dim localAddr As IPAddress = IPAddress.Loopback

            ' TcpListener server = new TcpListener(port);
            Dim server As TcpListener = New TcpListener(localAddr, port)

            ' Start listening for client requests.
            server.Start()

            ' Buffer for reading data
            Dim bytes(256) As Byte
            Dim data As String = Nothing

            ' Enter the listening loop.
            While True
                Console.Write("Waiting for a connection... ")

                ' Perform a blocking call to accept requests.
                ' You could also user server.AcceptSocket() here.
                Dim client As TcpClient = server.AcceptTcpClient()
                Console.WriteLine("Connected!")

                data = Nothing

                ' Get a stream object for reading and writing
                ' Wrap it in a NegotiateStream.
                negotiateStream = New NegotiateStream(client.GetStream())
                negotiateStream.AuthenticateAsServer()

                If negotiateStream.IsAuthenticated Then
                    Console.WriteLine( _
                        "IsAuthenticated: {0}", _
                        negotiateStream.IsAuthenticated)
                    Console.WriteLine( _
                        "IsMutuallyAuthenticated: {0}", _
                        negotiateStream.IsMutuallyAuthenticated)
                    Console.WriteLine( _
                        "IsEncrypted: {0}", _
                        negotiateStream.IsEncrypted)
                    Console.WriteLine( _
                        "IsSigned: {0}", _
                        negotiateStream.IsSigned)
                    Console.WriteLine( _
                        "IsServer: {0}", _
                        negotiateStream.IsServer)
                    Dim remoteIdentity As IIdentity = _
                         negotiateStream.RemoteIdentity
                    Console.WriteLine( _
                        "Client identity: {0}", _
                        remoteIdentity.Name)
                    Console.WriteLine( _
                        "Authentication Type: {0}", _
                        remoteIdentity.AuthenticationType)
                End If

                ' Loop to receive all the data sent by the client.
                Dim bytesRead As Integer = negotiateStream.Read(bytes, 0, bytes.Length)
                While bytesRead <> 0
                    ' Translate data bytes to a ASCII string.
                    data = _
                        System.Text.Encoding.ASCII.GetString(bytes, 0, bytesRead)
                    Console.WriteLine("Received: {0}", data)

                    ' Process the data sent by the client.
                    data = data.ToUpper( _
                        System.Globalization.CultureInfo.CurrentCulture)

                    Dim msg() As Byte = System.Text.Encoding.ASCII.GetBytes(data)

                    ' Send back a response.
                    negotiateStream.Write(msg, 0, msg.Length)
                    Console.WriteLine("Sent: {0}", data)

                    bytesRead = negotiateStream.Read(bytes, 0, bytes.Length)
                End While
            End While
        Catch ex As AuthenticationException
            Console.WriteLine(ex.Message)
        Catch ex As SocketException
            Console.WriteLine(ex.Message)
        Catch ex As IOException
            Console.WriteLine(ex.Message)
        Finally
            If negotiateStream IsNot Nothing Then
                negotiateStream.Close()
            End If
        End Try

        Console.WriteLine(Environment.NewLine & "Hit enter to continue...")
        Console.Read()
    End Sub
End Module
